packages/@aws-cdk/aws-iam/lib/grouping-by-actions-principal.ts

```diff
@@ -0,0 +1,68 @@
+import * as crypto from 'crypto';
+import { IIdentity } from './identity-base';
+import { PolicyStatement } from './policy-statement';
+import { AddToPrincipalPolicyResult, IGrantable, IPrincipal, PrincipalPolicyFragment } from './principals';
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct } from '@aws-cdk/core';
+
+/**
+ * An IAM Principal that wraps a different Principal,
+ * and groups multiple Policy Statements that share the same 'actions' property together,
+ * thus compressing the overall size of the Policy attached the given Principal,
+ * and making it less likely it hits the 10240 bytes IAM limit.
+ */
+export class GroupingByActionsPrincipal extends Construct implements IPrincipal, IGrantable {
+  public readonly assumeRoleAction: string;
+  public readonly policyFragment: PrincipalPolicyFragment;
+  public readonly principalAccount?: string | undefined;
+  public readonly grantPrincipal: IPrincipal;
+
+  /** The original principal that this class wraps. */
+  public readonly wrappedIdentity: IIdentity;
+  private readonly statements: { [key: string]: { statement: PolicyStatement, grantResult: AddToPrincipalPolicyResult } };
+
+  constructor(grantPrincipal: IIdentity, id: string) {
+    super(grantPrincipal as unknown as Construct, id);
+
+    this.assumeRoleAction = grantPrincipal.assumeRoleAction;
+    this.policyFragment = grantPrincipal.policyFragment;
+    this.principalAccount = grantPrincipal.principalAccount;
+    this.grantPrincipal = this;
+
+    this.wrappedIdentity = grantPrincipal;
+    this.statements = {};
+  }
+
+  public addToPolicy(statement: PolicyStatement): boolean {
+    return this.addToPrincipalPolicy(statement).statementAdded;
+  }
+
+  public addToPrincipalPolicy(statement: PolicyStatement): AddToPrincipalPolicyResult {
+    const key = this.keyFor(statement);
+    const statementCache = this.statements[key];
+    if (!statementCache) {
+      const grantResult = this.wrappedIdentity.addToPrincipalPolicy(statement);
+      this.statements[key] = { statement, grantResult };
+      return grantResult;
+    } else {
+      statementCache.statement.addResources(...statement.resources);
```

Comment:
`!` Since `statementCache.statement` is the same object as the original `PolicyStatement` given to you by the consumer, mutation will have unintended side effects:

```ts
const stmt = new PolicyStatement({ actions: ['s3:GetBanana'], resources: ['mybucket'] });

groupingByActionsPrincipal.addToPrincipalPolicy(stmt);
groupingByActionsPrincipal.addToPrincipalPolicy({ actions: ['s3:GetBanana'], resources: ['*'] });

stmt.addAction('aws:StealCredentials');  // Oopsepoops! End result is 'aws:StealCredentials' on '*' !
```

Only safe way around this is by making a copy of the incoming `PolicyStatement` to prevent the mutable aliasing.

That may break the legitimate use cases of mutating the original policy after the fact...

So probably the only way to do this correctly is to do the merging at render time.